experimental/lib/Transforms/Speculation/PlacementFinder.cpp

```diff
@@ -0,0 +1,359 @@
+
+// Recursively traverse the IR until reaching branches and store visited values
+static void markSpeculativePaths(Operation *currOp,
+                                 DenseSet<Value> &specValues) {
+  // End traversal when reaching a branch, because save units are only
+  // placed inside the speculation BB
+  if (isa<handshake::ConditionalBranchOp>(currOp))
+    return;
+
+  for (OpResult res : currOp->getResults()) {
+    if (specValues.contains(res))
+      continue;
+    specValues.insert(res);
+    for (Operation *succOp : res.getUsers()) {
+      markSpeculativePaths(succOp, specValues);
+    }
+  }
+}
+
+// Save units are needed where speculative tokens can interact with
+// non-speculative tokens
+LogicalResult PlacementFinder::findSavePositions() {
+  OpPlacement specPos = placements.getSpeculatorPlacement();
+  handshake::FuncOp funcOp =
+      specPos.dstOp->getParentOfType<handshake::FuncOp>();
+  assert(funcOp && "op should have parent function");
+  auto handshakeBlocks = getLogicBBs(funcOp);
+
+  // Mark all values that are speculative in the speculation BB
+  llvm::DenseSet<Value> specValues;
+  specValues.insert(specPos.srcOpResult);
+  markSpeculativePaths(specPos.dstOp, specValues);
+
+  // Iterate all operations in the speculation BB
+  unsigned bb = getLogicBB(specPos.dstOp).value();
+  for (Operation *blockOp : handshakeBlocks.blocks.lookup(bb)) {
+    // Create a save if an operation has both spec and non-spec operands
+    bool hasNonSpecInput = false;
+    bool hasSpecInput = false;
+    for (Value operand : blockOp->getOperands()) {
+      if (specValues.count(operand))
+        hasSpecInput = true;
+      else
+        hasNonSpecInput = true;
+    }
+
+    if (hasSpecInput and hasNonSpecInput) {
+      for (Value operand : blockOp->getOperands()) {
+        // Create a Save for every non-speculative operand
+        if (not specValues.count(operand)) {
+          // No save needed in front of Source Operations
+          if (isa<handshake::SourceOp>(operand.getDefiningOp()))
+            continue;
+
+          placements.addSave(operand, blockOp);
+        }
+      }
+    }
+  }
+
+  return success();
+}
+
+//===----------------------------------------------------------------------===//
+// Commit Units Finder Methods
+//===----------------------------------------------------------------------===//
+
+void PlacementFinder::findCommitsTraversal(llvm::DenseSet<Operation *> &visited,
+                                           Operation *currOp) {
+  // End traversal if currOp is already in visited set
+  if (auto [_, isNewOp] = visited.insert(currOp); !isNewOp)
+    return;
+  for (Value res : currOp->getResults()) {
+    for (Operation *succOp : res.getUsers()) {
+      if (placements.containsSave(res, succOp)) {
+        // A Commit is needed in front of Save Operations. This will be later
+        // converted into a SaveCommit for multiple loop speculation.
+        placements.addCommit(res, succOp);
```

Comment:
I know this is what is described in the paper but can't we just remove the save placement now and add a save commit placement?

Reply:
Done.